runtime/signals.c

```diff
@@ -287,43 +290,68 @@ void caml_request_minor_gc (void)
   caml_set_action_pending();
 }
 
-void caml_do_urgent_gc_and_callbacks(void)
+value caml_do_urgent_gc_and_callbacks_exn(void)
 {
+  value exn;
+
   caml_something_to_do = 0;
+
+  // Do any pending minor collection or major slice
   caml_check_urgent_gc(Val_unit);
+
   caml_update_young_limit();
-  caml_memprof_handle_postponed();
-  caml_final_do_calls();
-  caml_process_pending_signals();
+
+  // Call signal handlers first
+  exn = caml_process_pending_signals_exn();
+  if (Is_exception_result(exn)) goto exception;
+
+  // Call memprof callbacks
+  exn = caml_memprof_handle_postponed_exn();
+  if (Is_exception_result(exn)) goto exception;
+
+  // Call finalisers
+  exn = caml_final_do_calls_exn();
+  if (Is_exception_result(exn)) goto exception;
+
+  return Val_unit;
+
+exception:
+  /* If an exception is raised during an asynchronous callback, then
+     it might be the case that we did not run all the callbacks we
+     needed. Therefore, we set [caml_something_to_do] again in order
+     to force reexamination of callbacks. */
+  caml_set_action_pending();
+  return exn;
 }
 
 CAMLno_tsan /* The access to [caml_something_to_do] is not synchronized. */
-value caml_check_urgent_gc_and_callbacks(value extra_root)
+static inline value caml_check_urgent_gc_and_callbacks_exn(value extra_root)
 {
   if (caml_something_to_do) {
     CAMLparam1(extra_root);
-    caml_do_urgent_gc_and_callbacks();
+    value exn = caml_do_urgent_gc_and_callbacks_exn();
+    if (Is_exception_result(exn))
+      CAMLreturn(exn);
     CAMLdrop;
   }
   return extra_root;
 }
 
-/* If an exception is raised during an asynchronous callback (i.e.,
-   from an OCaml function called in `caml_check_urgent_gc`), then it
-   might be the case that we did not run all the callbacks we needed
-   to run even though [caml_something_to_do] has been reset to 0 at
-   the begining of [caml_check_urgent_gc]. Therefore, we set
-   [caml_something_to_do] in order to force reexamination of
-   callbacks.
-
-   Apart from a reasonable performance penalty (an extra call to
-   `caml_check_urgent_gc`), it is OK to call this function where
-   `caml_raise` would have been more appropriate (i.e., not called
-   from `caml_check_urgent_gc`). */
-void caml_raise_in_async_callback (value exc)
+value caml_check_urgent_gc_and_callbacks(value extra_root)
 {
-  caml_set_action_pending();
-  caml_raise(exc);
+  value res = caml_check_urgent_gc_and_callbacks_exn(extra_root);
+  return caml_raise_if_exception(res);
+}
+
+CAMLexport value caml_process_pending_actions_exn(void)
+{
+  return caml_check_urgent_gc_and_callbacks_exn(Val_unit);
+}
+
+CAMLexport void caml_process_pending_actions(void)
+{
+  value exn = caml_check_urgent_gc_and_callbacks_exn(Val_unit);
+  caml_raise_if_exception(exn);
 }
```

Comment:
I'm sorry, but I don't get the point of having these two functions, which are (appart from the extra root), aliases of `caml_check_urgent_gc_and_callbacks{_exn}`. Could you choose an API (either with an extra root or without), and have only one version of these functions?

Reply:
For deciding on a public interface, I adopt the point of view of the user. What they see is that `caml/signals.h` grows from:
```c
CAMLextern void caml_enter_blocking_section (void);
CAMLextern void caml_leave_blocking_section (void);
```
to:
```c
CAMLextern void caml_enter_blocking_section (void);
CAMLextern void caml_leave_blocking_section (void);

CAMLextern void caml_process_pending_actions (void);
/* Checks for pending actions and executes them. This includes pending
   minor and major collections, signal handlers, finalisers, and
   Memprof callbacks. Assumes that the runtime lock is held. Can raise
   exceptions asynchronously into OCaml code. */

CAMLextern value caml_process_pending_actions_exn (void);
/* Same as [caml_process_pending_actions], but returns the exception
   if any (otherwise returns [Val_unit]). */
```
They do not care that for internal and historical reasons we ended up with `caml_check_urgent_gc_and_callbacks` as it is today. Besides the name which I choose to be a bit better since it is a blank slate (from the user POV), I think there is another important difference there: `caml_check_urgent_gc_and_callbacks` takes an extra root because in many of the primitives where it is used, the values are not rooted and we need only one root momentarily. Maybe this was some form of optimisation. But this is not what is advocated by the FFI documentation, and if the user follows the manual (Section 20.6) then they will never need an extra root like that.

Once we've decided for a public API that does not take an extra root, we can ask, can we simplify the internals using the public API, and remove `caml_check_urgent_gc_and_callbacks`? This is a natural question, and it is debatable. This concerns a few `array.c` primitives, `ephemeron_create` in `weak.c`, and `intern_end` in `intern.c` where I have to add roots manually and unconditionally. Remember, I do not know if there are performance implications; by default I assume that they are written in this special style for a reason. So instead, I used Stephen's trick to avoid rooting the value when not needed.

Since you agree with the technical soundness, to move forward, we now need to agree on the public interface. I propose it as above. Anybody interested can chime in.

Once we agree on the public interface, I am interested in hearing your nitpicks (for instance I can remove `caml_check_urgent_gc_and_callbacks{,_exn}` if you tell me that there are no performance implications).